### PR TITLE
per tenant retention

### PR DESCRIPTION
**Pull request by jnyi**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/compact/retention.go [hunk 1]
```diff
@@ -47,3 +53,75 @@ func ApplyRetentionPolicyByResolution(
...
+			policy.RetentionDuration = time.Duration(duration)
+		}
+		level.Info(logger).Log("msg", "retention policy for tenant is enabled", "tenant", tenant, "retention policy", policy)
+		retentionByTenant[tenant] = policy
+	}
+	return retentionByTenant, nil
+}
+
+// ApplyRetentionPolicyByTenant removes blocks depending on the specified retentionByTenant based on blocks MaxTime.
+func ApplyRetentionPolicyByTenant(
```
Comment (jnyi) [inline pkg/compact/retention.go:106]: plan to add unit test for this as well

File: cmd/thanos/compact.go [hunk 2]
```diff
@@ -436,6 +436,12 @@ func runCompact(
 		level.Info(logger).Log("msg", "retention policy of 1 hour aggregated samples is enabled", "duration", retentionByResolution[compact.ResolutionLevel1h])
 	}
 
+	retentionByTenant, err := compact.ParesRetentionPolicyByTenant(logger, *conf.retentionTenants)
+	if err != nil && len(retentionByTenant) != 0 {
```
Comment (hczhu-db) [inline cmd/thanos/compact.go]: Why is `len(retentionByTenant) != 0` needed here? If it's an empty config, `err` should already be nil.

File: pkg/compact/retention.go [hunk 3]
```diff
@@ -6,19 +6,25 @@ package compact
...
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
 	"github.com/thanos-io/objstore"
 
 	"github.com/thanos-io/thanos/pkg/block"
 	"github.com/thanos-io/thanos/pkg/block/metadata"
 )
 
+const (
+	tenantRetentionRegex = `^([\w-]+):((\d{4}-\d{2}-\d{2})|(\d+d))$`
```
Comment (hczhu-db) [inline pkg/compact/retention.go:27]: Give some examples in a piece of code comment.

File: pkg/compact/retention.go [hunk 4]
```diff
@@ -47,3 +53,75 @@ func ApplyRetentionPolicyByResolution(
...
+		tenant := matches[1]
+		var policy RetentionPolicy
+		if _, ok := retentionByTenant[tenant]; ok {
+			return nil, errors.Errorf("duplicate retention policy for tenant: %s", tenant)
+		}
+		if cutoffDate, err := time.Parse(time.DateOnly, matches[3]); err != nil && matches[3] != "" {
+			return nil, errors.Wrapf(invalidFormat, "error parsing cutoff date: %v", err)
+		} else if matches[3] != "" {
+			policy.CutoffDate = cutoffDate
+		}
```
Comment (hczhu-db) [inline pkg/compact/retention.go]: nit: don't need `else if{}` because the function returns in `if{}`.

File: pkg/compact/retention.go [hunk 5]
```diff
@@ -47,3 +53,75 @@ func ApplyRetentionPolicyByResolution(
...
+		if duration, err := model.ParseDuration(matches[4]); err != nil && matches[4] != "" {
+			return nil, errors.Wrapf(invalidFormat, "error parsing duration: %v", err)
+		} else if matches[4] != "" {
+			policy.RetentionDuration = time.Duration(duration)
+		}
```
Comment (hczhu-db) [inline pkg/compact/retention.go]: ditto

**Review by hczhu-db — APPROVED** [review]

File: cmd/thanos/compact.go (hunk 2 above)
Comment (jnyi) [reply cmd/thanos/compact.go]: good catch

File: pkg/compact/retention.go (hunk 4 above)
Comment (jnyi) [reply pkg/compact/retention.go]: still need this, `matches[3]` can be empty, we allow optional cutoffDate or a retention duration

**Review by yuchen-db — APPROVED** [review]
lgtm

File: pkg/compact/retention.go (hunk 3 above)
Comment (jnyi) [reply pkg/compact/retention.go:27]: examples in unit tests already

**Pull request merged**